### PR TITLE
[quant][graphmode] Swap quantized functional linear with aten::linear

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33650 [not4land] debugging
* #33533 [not4land] Temporary workaround to get quantized ops
* #34414 [quant][graphmode][refactor] insertObservers for Block
* #34411 [quant][graphmode][fix] Insert dequantize before use node
* #34349 [quant][graphmode][fix] preserve the type of original value when inserting dequant node
* #34348 [quant] Add dequantize.tensors
* #34347 [quant][graphmode] Add quantization support for aten::dropout
* #34346 [quant][graphmode] Add quantization support for aten::cat
* #34345 [quant][graphmode] Add prim::ListConstruct to general op handling
* #34315 [quant][graphmode] insert quant/dequant work for duplicated debugName
* #34314 [quant][graphmode][refactor] Change QParamMap to QParamVector
* #33895 [quant][graphmode] insert prepack/unpack for aten::linear
* #33854 [quant][graphmode] Quantization pattern for aten::linear
* #33532 [quant][graphmode] Add quantization pattern for quantized::add_relu
* #33279 [quant][graphmode] Add quantized conv2d-relu fusion pattern
* #33927 [quant][graphmode] Add Finalize function that inlines graph and produce quantized ops
* **#33853 [quant][graphmode] Swap quantized functional linear with aten::linear**
* #33481 [quant][graphmode] Handling ops doesn't require observation in insertObservers

Summary:
Quant fusion relies on inline, but inline will break the CallFunction("linaer", ...) into a if block
it will be hard to recognize this block and swap it with quantized::linear, in order to
preserve the op, we will swap all quantized functional linear into aten::linear.
They might produce different backward graph, but this is called in the step before we get quantized
model, so it shouldn't affect anything.
We'll integrate this with convert_script later in the new "finalize_quant" API

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20343873](https://our.internmc.facebook.com/intern/diff/D20343873)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -1911,6 +1911,34 @@ def forward(self, x):
...
+        bias = torch.rand((5), dtype=torch.float)
+        m = torch.jit.script(M())
+        ref_res = m(x, weight, bias)
+        FileCheck().check("CallFunction") \
+                   .run(m.graph)
+        torch._C._jit_pass_swap_functional_linear(m.graph)
+        FileCheck().check("aten::linear") \
+                   .check_not("CallFunction") \
+                   .run(m.graph)
+        res = get_forward(m._c)(x, weight, bias)
```
Comment (raghuramank100) [inline test/test_jit.py]: Can one also do res = m(x, weight, bias)? Why is get_forward needed?

File: torch/csrc/jit/passes/quantization.cpp [hunk 2]
```diff
@@ -2007,6 +2007,39 @@ void FoldQuantNodesIntoInputsOutputs(std::shared_ptr<Graph>& graph) {
...
+                   const std::unordered_map<std::string, Value*>& vmap) {
+    const auto& match_vmap = match.values_map;
+    auto linear = getValue("linear", match_vmap, vmap);
+    auto func_name = getFuncName(linear);
+    return func_name == "linear";
+  };
+  SubgraphRewriter rewriter;
+  rewriter.RegisterRewritePattern(functional_linear, aten_linear);
+  // TODO: runOnGraph takes const ref?
+  rewriter.runOnGraph(graph);
```
Comment (raghuramank100) [inline torch/csrc/jit/passes/quantization.cpp]: Missing argument filter?

File: torch/csrc/jit/python/init.cpp [hunk 3]
```diff
@@ -213,6 +213,8 @@ void initJITBindings(PyObject* module) {
       .def("_jit_pass_dedup_module_uses", &DedupModuleUses)
       .def("_jit_pass_replicate_dequantize", &ReplicateDeQuant)
       .def("_jit_pass_swap_dequantize", &SwapDeQuant)
+      .def("_jit_pass_swap_functional_linear", &SwapFunctionalLinear)
+      .def("_jit_pass_swap_functional_linear_in_module", &SwapFunctionalLinearInModule)
```
Comment (raghuramank100) [inline torch/csrc/jit/python/init.cpp]: I assume this pass is invoked prior to quant-fusion. Can you also make a change to convert_script to include this pass?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see comments

File: test/test_jit.py (hunk 1 above)
Comment (jerryzh168) [reply test/test_jit.py]: yeah, I think we probably don't need geet_forward, I'll change this

File: torch/csrc/jit/python/init.cpp (hunk 3 above)
Comment (jerryzh168) [reply torch/csrc/jit/python/init.cpp]: This is called in Finalize, see future PRs in the stack

**Review by raghuramank100 — APPROVED** [review]

**Pull request closed without merging**